FreeRTOS-Plus/Source/FreeRTOS-IoT-Libraries-LTS-Beta2/c_sdk/platform/freertos/retry_utils/retry_utils_freertos.c

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright (C) 2020 Amazon.com, Inc. or its affiliates.  All Rights Reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/**
+ * @file retry_utils_freertos.c
+ * @brief Utility implementation of backoff logic, used for attempting retries of failed processes.
+ */
+
+/* Standard includes. */
+#include <stdint.h>
+
+/* Kernel includes. */
+#include "FreeRTOS.h"
+#include "task.h"
+
+#include "retry_utils.h"
+
+#define _MILLISECONDS_PER_SECOND    ( 1000U )                                                         /**< @brief Milliseconds per second. */
+
+extern UBaseType_t uxRand( void );
+
+/*-----------------------------------------------------------*/
+
+RetryUtilsStatus_t RetryUtils_BackoffAndSleep( RetryUtilsParams_t * pRetryParams )
+{
+    RetryUtilsStatus_t status = RetryUtilsRetriesExhausted;
+    int32_t backOffDelayMs = 0;
+
+    /* If MAX_RETRY_ATTEMPTS is set to 0, try forever. */
+    if( ( pRetryParams->attemptsDone < MAX_RETRY_ATTEMPTS ) ||
+        ( 0 == MAX_RETRY_ATTEMPTS ) )
+    {
+        /* Choose a random value for back-off time between 0 and the max jitter value. */
+        backOffDelayMs = uxRand() % pRetryParams->nextJitterMax;
+
+        /*  Wait for backoff time to expire for the next retry. */
+        vTaskDelay( pdMS_TO_TICKS( backOffDelayMs * _MILLISECONDS_PER_SECOND ) );

```

Comment:
Seem to be conflating the backoff time and the jitter.  To my mind, this function should sleep for backoff + [0, max jitter], not [0, backoff + jitter].

Reply:
While I do agree, I did not write this retry logic and am simply using the same implementation in CSDK. I am afraid that changing it might break something. Please feel free to chime in, @abhidixi11.

Reply:
No, I think it's correct here. Sleeping for backoff + [0, max jitter] is closer to the "Equal jitter" approach described [here](https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/), while the current implementation is closer to "Full jitter". In the blog, the equal jitter approach lost to full jitter in an experiment, in both completion time and work done.

Reply:
Interesting article but how well do these simulations apply to things outside OCC? In the case of connecting to a server vs.  writing a database transaction, multiple clients would be able to succeed in the same round. It would be nice if the application could define the backoff algorithm (maybe for the well-architected SDK?). I imagine that some types of jitter work better for certain use cases than others.

Reply:
I'm also a bit skeptical about applying the conclusions of the article to our use cases.  Still, I think these are reasonable arguments for keeping this logic as is.

Reply:
Idk, but it is the article that we ourselves link to in this same file, so I thought it'd have to be at least somewhat relevant.